HW_3/hw_3.py

```diff
@@ -0,0 +1,308 @@
+2. Реализовать предыдущий пункт, не создавая файлов на диске (**+2 дополнительных балла**)
+3. Если функция ничего не печатает в `stdout` и `stderr` &mdash; отправлять файл не нужно
+
+**Важные примечания:**
+1. Ни в коем случае не храните свой API токен в коде и не загружайте его ни в каком виде свой в репозиторий. Сохраните его в **переменной окружения** `TG_API_TOKEN`, тогда его можно будет получить из кода при помощи `os.getenv("TG_API_TOKEN")`. Ручное создание переменных окружения может быть не очень удобным, поэтому можете воспользоваться функцией `load_dotenv` из модуля [dotenv](https://pypi.org/project/python-dotenv/). В доке всё написано, но если коротко, то нужно создать файл `.env` в текущей папке и записать туда `TG_API_TOKEN=<your_token>`, тогда вызов `load_dotenv()` создаст переменные окружения из всех переменных в файле. Это довольно часто используемый способ хранения ключей и прочих приватных данных
+2. Функцию `long_lasting_function` из примера по понятным причинам запускать не нужно. Достаточно просто убедится, что большие временные интервалы правильно форматируются при отправке сообщения (как в примерах)
+3. Допустима реализация логирования, когда логгер полностью перехватывает запись в `stdout` и `stderr` (то есть при выполнении функций печать происходит **только** в файл)
+4. В реальной жизни вам не нужно использовать Telegram API при помощи ручных запросов, вместо этого стоит всегда использовать специальные библиотеки Python, реализующие Telegram API, они более высокоуровневые и удобные. В данном задании мы просто учимся работать с API при помощи написания велосипеда.
+5. Обязательно прочтите часть конспекта лекции про API перед выполнением задания, так как мы довольно поверхностно затронули это на лекции
+
+**Рекомендуемые к использованию модули:**
+1. os
+2. sys
+3. io
+4. datetime
+5. requests
+6. dotenv
+
+**Запрещённые модули**:
+1. Любые библиотеки, реализующие Telegram API в Python (*python-telegram-bot, Telethon, pyrogram, aiogram, telebot* и так далле...)
+2. Библиотеки, занимающиеся "перехватыванием" данных из `stdout` и `stderr` (*pytest-capturelog, contextlib, logging*  и так далле...)
+
+
+
+Результат запуска кода ниже должен быть примерно такой:
+
+![image.png](attachment:620850d6-6407-4e00-8e43-5f563803d7a5.png)
+
+![image.png](attachment:65271777-1100-44a5-bdd2-bcd19a6f50a5.png)
+
+![image.png](attachment:e423686d-5666-4d81-8890-41c3e7b53e43.png)
+"""
+
+import sys
+import time
+
+telegram_bot_token = os.getenv('TELEGRAM_BOT_TOKEN')
+telegram_chat_id = os.getenv('TELEGRAM_CHAT_ID')
+
+def telegram_logger(chat_id):
+
+    def wrapper(func):
+        def logger_function(*args, **kwargs):
+            prev_stdout = sys.stdout
+            prev_stderr = sys.stderr
+            out_buffer = StringIO()
+            err_buffer = StringIO()
+            sys.stdout = out_buffer
+            sys.stderr = err_buffer
+            start_time = time.monotonic()
```

Comment:
Объекты `datetime` можно вычитать, в результате сразу получится `timedelta`
```suggestion
            start_time = datetime.now()
```